**Fill the Code blocks according to the comments provided as hints**

In [2]:
from sklearn.datasets import load_diabetes #We will use this dataset
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split
#import libraries for numpy, Linear regression, r2_score, train_test_split

In [3]:
X,y = load_diabetes(return_X_y=True)  #loading dataset
diab = load_diabetes()
print(diab.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - age     age in years
      - sex
      - bmi     body mass index
      - bp      average blood pressure
      - s1      tc, total serum cholesterol
      - s2      ldl, low-density lipoproteins
      - s3      hdl, high-density lipoproteins
      - s4      tch, total cholesterol / HDL
      - s5      ltg, possibly log of serum triglycerides level
      - s6      glu, blood sugar level

Note: Each of these 1

In [4]:
#print shape of X and y and write in a text block what do you observe about both 
print(X.shape)
print(y.shape)

(442, 10)
(442,)


X is the data of load_datasets and it has 442 ros and 10 columns. This means that load_datasets has 10 attributes and 442 entries
Y is the target of load_dataseta i.e what we try to determine using this dataset which is to get a quantitative measure of disease progression one year after baseline. Thus, it has only 442 elements corresponding to the 442 entries.
We can say that target is the 11th column in the dataset which determines the measure of disease progression over one year.

In [5]:
#split the dataset into Train(80%) and Test(20%)

X_train, X_test, y_train, y_test = train_test_split(X,y, train_size = 0.8, random_state = 42)
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(353, 10)
(89, 10)
(353,)
(89,)


In [6]:
#fit X,y in Linear Regression Model of sklearn
reg = LinearRegression()
reg.fit(X_train,y_train)


LinearRegression()

In [7]:
#print coefficients and intercepts (in @d gradient descent in was only m and b but as it is n dimensional it will have n coefficients and 1 intercept)
print(reg.coef_)
print(reg.intercept_)



[  37.90402135 -241.96436231  542.42875852  347.70384391 -931.48884588
  518.06227698  163.41998299  275.31790158  736.1988589    48.67065743]
151.34560453985995


In [8]:
#predict on X_test and calculate r2_score
y_test_pred = reg.predict(X_test)
print(y_test_pred, "\n\n")
r2 = round(r2_score(y_test, y_test_pred),3)
print(r2)



[139.5475584  179.51720835 134.03875572 291.41702925 123.78965872
  92.1723465  258.23238899 181.33732057  90.22411311 108.63375858
  94.13865744 168.43486358  53.5047888  206.63081659 100.12925869
 130.66657085 219.53071499 250.7803234  196.3688346  218.57511815
 207.35050182  88.48340941  70.43285917 188.95914235 154.8868162
 159.36170122 188.31263363 180.39094033  47.99046561 108.97453871
 174.77897633  86.36406656 132.95761215 184.53819483 173.83220911
 190.35858492 124.4156176  119.65110656 147.95168682  59.05405241
  71.62331856 107.68284704 165.45365458 155.00975931 171.04799096
  61.45761356  71.66672581 114.96732206  51.57975523 167.57599528
 152.52291955  62.95568515 103.49741722 109.20751489 175.64118426
 154.60296242  94.41704366 210.74209145 120.2566205   77.61585399
 187.93203995 206.49337474 140.63167076 105.59678023 130.70432536
 202.18534537 171.13039501 164.91423047 124.72472569 144.81030894
 181.99635452 199.41369642 234.21436188 145.95665512  79.86703276
 157.369412

In [9]:
#print X_train shape
print(X_train.shape)
##print(y_train.reshape(353,1))

x=np.array([1,2,3,4,5,6,7,8,9,10])
##y=np.array([2,3,4,5,6,7,8,9,10,11])
print(x+5)
y_hat = np.matmul(np.ones((1,10)), np.transpose(X_train)) + 4
print(2* np.sum(y_hat.ravel() - y_train))

(353, 10)
[ 6  7  8  9 10 11 12 13 14 15]
-105709.99426683718


In [11]:
 #Write GDregressor class


class GDRegressor:
    
    def __init__(self,learning_rate,epochs):
        
        self.coef_ = None
        self.intercept_ = None
        self.lr = learning_rate
        self.epochs = epochs
        
    def fit(self,X_train,y_train):
        # init your coefs
        self.intercept_ = 0 #initialise the intercept
        self.coef_ = np.ones((1,10))#initialise the array of coefficients according to number of coefficients
        
        for i in range(self.epochs):
            # update all the coef and the intercept
            
            y_hat = np.matmul(self.coef_, np.transpose(X_train)) + self.intercept_ #apply prediction formula i.e. (y_hat = coefficient*X + intercept) in terms of matrices coefficient and X   
            
            
            intercept_derivative = self.lr* np.sum(np.matmul(self.coef_, np.transpose(X_train)) + self.intercept_  - y_train) #find derivative matrix 
            self.intercept_ = self.intercept_ - intercept_derivative #update intercept using above intercept_derivative
            
            coef_derivative = self.lr*np.matmul((np.matmul(self.coef_, np.transpose(X_train)) + self.intercept_  - y_train).reshape(1,353), X_train)  #derivative matrix for all coefficients in a single step
            self.coef_ = self.coef_ - coef_derivative #update intercept using coef_derivative
        print(self.coef_,"\n\n")
        print(self.intercept_)
        
        #print the coefficients array and the intercept
    
    def predict(self,X_test):
        return  (np.matmul(self.coef_, np.transpose(X_test)) + self.intercept_ ) #new prediction in the form of (y = coefficients*X + intercept) using matrices

In [12]:
#call GDRegressor by keeping epochs and learning rate according to you and keep changing it for better results
obj = GDRegressor(0.01,50)
obj.fit(X_train,y_train)

[[ 8.04765183e+19  1.04126185e+19  9.69635992e+19  6.59955539e+19
  -3.07089701e+19 -4.46630842e+19 -5.53459126e+19  2.13301411e+19
   6.80771076e+19  1.05627599e+20]] 


-2.1967888714943974e+22


In [13]:
#fit X_train, y_train into GDRegressor object using function of the class
obj.fit(X_train,y_train)

[[ 8.04765183e+19  1.04126185e+19  9.69635992e+19  6.59955539e+19
  -3.07089701e+19 -4.46630842e+19 -5.53459126e+19  2.13301411e+19
   6.80771076e+19  1.05627599e+20]] 


-2.1967888714943974e+22


In [14]:
#predict on X_test using function of the class
print(obj.predict(X_test).reshape(1,89))
r2 = round(r2_score(y_test, np.transpose(obj.predict(X_test))),3)
print(r2)

[[-2.19744668e+22 -2.19590916e+22 -2.19680809e+22 -2.19332420e+22
  -2.19722453e+22 -2.19836526e+22 -2.19454132e+22 -2.19585637e+22
  -2.19829603e+22 -2.19785948e+22 -2.19891790e+22 -2.19729066e+22
  -2.19951684e+22 -2.19455439e+22 -2.19701577e+22 -2.19651085e+22
  -2.19426557e+22 -2.19372500e+22 -2.19522859e+22 -2.19481055e+22
  -2.19660234e+22 -2.19861218e+22 -2.19870323e+22 -2.19526091e+22
  -2.19744590e+22 -2.19570998e+22 -2.19565802e+22 -2.19566061e+22
  -2.19963946e+22 -2.19703706e+22 -2.19618376e+22 -2.19911318e+22
  -2.19690554e+22 -2.19539623e+22 -2.19592849e+22 -2.19676389e+22
  -2.19679559e+22 -2.19699725e+22 -2.19544560e+22 -2.19926559e+22
  -2.19804660e+22 -2.19674144e+22 -2.19624663e+22 -2.19532500e+22
  -2.19632346e+22 -2.19962035e+22 -2.19889880e+22 -2.19746353e+22
  -2.19912246e+22 -2.19781779e+22 -2.19685760e+22 -2.19995807e+22
  -2.19751649e+22 -2.19762152e+22 -2.19731293e+22 -2.19758135e+22
  -2.19827479e+22 -2.19521197e+22 -2.19836860e+22 -2.19882697e+22
  -2.19692

In [ ]:
#calculate r2_score to verify
